### DISCUSSION TITLE
为什么 Redis 持久化使用子进程 · Why's THE Design? · /whys-the-design-redis-bgsave-fork

### DISCUSSION THREAD

**Discussion by draveness** [draveness/blog-comments#580] [Announcements]
https://draveness.me/whys-the-design-redis-bgsave-fork

**Comment by arrayadd** [comment]
java 中针对读多写少场景也提供了一种同名的机制，例如CopyOnWriteArrayList。但是看源码，写的时候是真正的复制出一个数组来，使用时候还真要注意实际场景,了解机制背后的“Why’s THE Design”非常重要，感谢楼主系列。
```java
public boolean add(E e) {
		//写时候加锁
        final ReentrantLock lock = this.lock;
        lock.lock();
        try {
            Object[] elements = getArray();
            int len = elements.length;
            //复制一份数据
            Object[] newElements = Arrays.copyOf(elements, len + 1);
            newElements[len] = e;
            setArray(newElements);
            return true;
        } finally {
            lock.unlock();
        }
    }
```

**Comment by draveness** [comment]
> java 中针对读多写少场景也提供了一种同名的机制，例如CopyOnWriteArrayList。但是看源码，写的时候是真正的复制出一个数组来，使用时候还真要注意实际场景,了解机制背后的“Why’s THE Design”非常重要，感谢楼主系列。

这确实会让人对『写时复制』造成误解，有歧义的命名可能会导致很多问题，感谢支持。

**Comment by yuanjize** [comment]
“共享内存”指的是哪些内存，是堆吗。

**Comment by draveness** [comment]
> “共享内存”指的是哪些内存，是堆吗。

不止包括堆，栈、数据段等都会共享

**Comment by leafduo** [comment]
第一次了解到 copy-on-write 是在文件系统里，例如 btrfs 等

**Comment by GavinCruise** [comment]
> @draveness
> > java 中针对读多写少场景也提供了一种同名的机制，例如CopyOnWriteArrayList。但是看源码，写的时候是真正的复制出一个数组来，使用时候还真要注意实际场景,了解机制背后的“Why’s THE Design”非常重要，感谢楼主系列。
> 
> 这确实会让人对『写时复制』造成误解，有歧义的命名可能会导致很多问题，感谢支持。

Java 里的 CopyOnWriteArrayList 属于 CoW 容器，更像是借用了写时复制思想为并发读做的一种解决方案，以 CopyOnWrite 命名也还行

**Comment by GavinCruise** [comment]
Docker 中从 image 构建 container 也是用到了 copy-on-write 机制。同样感谢博主这一系列文章，受益匪浅~

**Comment by rzzdy** [comment]
> @draveness
> > “共享内存”指的是哪些内存，是堆吗。
> 
> 不止包括堆，栈、数据段等都会共享

像输出缓冲区这种也是共享的么，fork后，子进程再继承一份？

**Comment by selboo** [comment]
相关文章里面的 
[013 为什么 UDP 头只有 8 个字节](https://draveness.me/whys-the-design-tcp-minimum-header)  连接不对, 打不开


应该是 whys-the-design-udp-minimum-header 吧

**Comment by draveness** [comment]
> @rzzdy
> > @draveness
> > > “共享内存”指的是哪些内存，是堆吗。
> > 
> > 不止包括堆，栈、数据段等都会共享
> 
> 像输出缓冲区这种也是共享的么，fork后，子进程再继承一份？

fork 确实会『复制』一份输出缓冲区，但是『再继承一份』是指？

**Comment by draveness** [comment]
> @selboo
> 相关文章里面的 
> [013 为什么 UDP 头只有 8 个字节](https://draveness.me/whys-the-design-tcp-minimum-header)  连接不对, 打不开
> 
> 
> 应该是 whys-the-design-udp-minimum-header 吧

感谢，已经在发布了

**Comment by vampire007** [comment]
首先感谢博主分享~ 本文中您提到了 rdb 快照机制，能否再介绍下redis aof 的机制呢？

**Comment by draveness** [comment]
> 首先感谢博主分享~ 本文中您提到了 rdb 快照机制，能否再介绍下redis aof 的机制呢？

AOF 其实不是这篇文章的想要回答的问题，你有什么关于 AOF 的问题么？

**Comment by Coolxc** [comment]
请问RDB中fork的子进程是不是就永远不会复制内存了，因为生成快照并且保存到磁盘这一过程不会出现写操作

**Comment by draveness** [comment]
> 请问RDB中fork的子进程是不是就永远不会复制内存了，因为生成快照并且保存到磁盘这一过程不会出现写操作

父进程可能会写入，这是会触发内存页的复制

**Comment by xing393939** [comment]
为什么不能另外起一个线程而是子进程来快照呢？

**Reply by LazyPigPig** [reply]
多线程共享进程的虚拟内存空间，主线程修改内存的时候，rdb线程会感知到内存变化。

**Comment by draveness** [comment]
> 为什么不能另外起一个线程而是子进程来快照呢？

先看看进程和线程的区别

**Comment by OrdinaryYZH** [comment]
> @draveness
> > 为什么不能另外起一个线程而是子进程来快照呢？
> 
> 先看看进程和线程的区别

个人理解是：可以，但是吞吐量会降低，没必要。

我也好奇LZ的解释是什么呢？

**Comment by OrdinaryYZH** [comment]
> @draveness
> > 首先感谢博主分享~ 本文中您提到了 rdb 快照机制，能否再介绍下redis aof 的机制呢？
> 
> AOF 其实不是这篇文章的想要回答的问题，你有什么关于 AOF 的问题么？

AOF的重写，也会使用了fork，所以文章中的内容还是对理解aof的重写有作用的，参考：https://redisbook.readthedocs.io/en/latest/internal/aof.html#id13

**Comment by draveness** [comment]
> > @draveness
> > > 为什么不能另外起一个线程而是子进程来快照呢？
> > 
> > 先看看进程和线程的区别
> 
> 个人理解是：可以，但是吞吐量会降低，没必要。
> 
> 我也好奇LZ的解释是什么呢？

线程会共享内存可能还需要处理竞争条件的问题，进程有独立的内存空间，实现上会简单很多。

**Comment by MalcolmSanders** [comment]
> @xing393939
> 为什么不能另外起一个线程而是子进程来快照呢？

如果用线程来做快照，在做快照的过程中，所有写操作会立即对redis状态产生影响，这样最终做出来的快照整体上就不是全局一致的了，即可能出现一个快照中，有些部分的数据较新（在快照阶段后期持久化），有些数据版本更老（在快照阶段前期持久化）。

**Comment by OrdinaryYZH** [comment]
> @MalcolmSanders
> > @xing393939
> > 为什么不能另外起一个线程而是子进程来快照呢？
> 
> 如果用线程来做快照，在做快照的过程中，所有写操作会立即对redis状态产生影响，这样最终做出来的快照整体上就不是全局一致的了，即可能出现一个快照中，有些部分的数据较新（在快照阶段后期持久化），有些数据版本更老（在快照阶段前期持久化）。

想了下，好像是这个道理，因为cow，所以容易实现某一个时间的快照

**Comment by dayuange123** [comment]
redis 主从复制 需要rdb，但是怎么保证从节点加载rdb后。后续收到的命令恰好是这个rdb之后的命令呢？如果命令重复或者丢失 也会不一致把

**Comment by ghaoxiangzZ** [comment]
请问博主写时复制 是以页为单位全量复制吗？还是复制有数据变更的页，无变更的页另外处理

**Comment by BrianQy** [comment]
AOF和RDB的fork子进程的行为特点有什么不同嘛

**Comment by shuangquansq** [comment]
作者你好，文中这句话“也就是说 Redis 对机器上内存资源的最大利用率不能超过 50%”在写时复制场景下面似乎有问题，因为写时复制真正只有变化的内存页面才会被复制出来，不变的不会，那么父子进程使用的内存就不是双倍的关系。请问，是不是这样？
